File: c-lib/examples/tls.c
/* Self-contained TLS demo: a bind socket and a connect socket talking over
 * OpenSSL in one process.
 *
 *     ./tls [cert-dir]     # cert-dir defaults to ../conformance/certs
 *
 * It loads the repository's shared self-signed certificate (EC P-256, with
 * SANs for "localhost" and 127.0.0.1). A real deployment points cert/key at its
 * own PEM files and trusts the issuing CA via the ca path. The TCP target and
 * the verified name are independent: we dial 127.0.0.1 but verify "localhost". */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "aiomsg.h"

int main(int argc, char **argv) {
    const char *dir = argc > 1 ? argv[1] : "../conformance/certs";
    char cert[512], key[512];
    snprintf(cert, sizeof cert, "%s/cert.pem", dir);
    snprintf(key, sizeof key, "%s/key.pem", dir);

    aiomsg_socket *server = aiomsg_new(AIOMSG_PUBLISH, AIOMSG_AT_MOST_ONCE, NULL);
    if (aiomsg_bind_tls(server, "127.0.0.1", 25001, cert, key) != 0) {
        fprintf(stderr, "bind_tls failed (is %s readable?)\n", cert);
        return 1;
    }
    printf("server bound (TLS) on 127.0.0.1:25001\n");

    aiomsg_socket *client = aiomsg_new(AIOMSG_ROUNDROBIN, AIOMSG_AT_MOST_ONCE, NULL);
    if (aiomsg_connect_tls(client, "127.0.0.1", 25001, cert, "localhost") != 0) {
        fprintf(stderr, "connect_tls failed\n");
        return 1;
    }

    const char *msg = "hello over TLS";
    aiomsg_send(server, msg, strlen(msg));

    void *data;
    size_t len;
    if (aiomsg_recv_timeout(client, &data, &len, NULL, 5000) == 1) {
        printf("client received: %.*s\n", (int)len, (char *)data);
        free(data);
    } else {
        printf("timed out\n");
    }

    aiomsg_close(client);
    aiomsg_free(client);
    aiomsg_close(server);
    aiomsg_free(server);
    return 0;
}

File: rust-lib-async/examples/tls.rs
//! Self-contained TLS demo: a bind socket and a connect socket talking over
//! rustls in one process. Run with:
//!
//! ```text
//! cargo run --example tls
//! ```
//!
//! It generates a throwaway self-signed certificate at startup so it runs with
//! no setup. A real deployment loads its certificate chain and private key from
//! PEM files instead — see [`server_config_from_pem`] below for that shape.

use std::sync::Arc;

use aiomsg::rustls::crypto::ring;
use aiomsg::rustls::pki_types::{CertificateDer, PrivateKeyDer, PrivatePkcs8KeyDer};
use aiomsg::rustls::{ClientConfig, RootCertStore, ServerConfig};
use aiomsg::Socket;

#[tokio::main]
async fn main() -> aiomsg::Result<()> {
    let (server_cfg, client_cfg) = self_signed_configs("localhost");

    let server = Socket::new();
    let addr = server.bind_tls("127.0.0.1:0", server_cfg).await?;
    println!("server bound (TLS) on {addr}");

    let client = Socket::new();
    // The TCP target is `addr`, but the certificate is validated against the
    // name "localhost" — these are independent.
    client.connect_tls(addr, "localhost", client_cfg).await?;

    server.send("hello over TLS").await?;
    if let Some(msg) = client.recv().await {
        println!("client received: {}", String::from_utf8_lossy(&msg));
    }

    server.close().await;
    client.close().await;
    Ok(())
}

/// Generate a matching (server, client) config pair for `name`, trusting only
/// the freshly minted certificate. Handy for tests and demos.
fn self_signed_configs(name: &str) -> (Arc<ServerConfig>, Arc<ClientConfig>) {
    let ck = rcgen::generate_simple_self_signed(vec![name.to_string()]).unwrap();
    let cert = CertificateDer::from(ck.cert.der().to_vec());
    let key = PrivateKeyDer::Pkcs8(PrivatePkcs8KeyDer::from(ck.key_pair.serialize_der()));
    let provider = Arc::new(ring::default_provider());

    let server = ServerConfig::builder_with_provider(provider.clone())
        .with_safe_default_protocol_versions()
        .unwrap()
        .with_no_client_auth()
        .with_single_cert(vec![cert.clone()], key)
        .unwrap();

    let mut roots = RootCertStore::empty();
    roots.add(cert).unwrap();
    let client = ClientConfig::builder_with_provider(provider)
        .with_safe_default_protocol_versions()
        .unwrap()
        .with_root_certificates(roots)
        .with_no_client_auth();

    (Arc::new(server), Arc::new(client))
}

/// How a real server builds its config: read the certificate chain and private
/// key from PEM files (e.g. from your CA or Let's Encrypt). Unused by the demo;
/// shown here as the realistic counterpart to [`self_signed_configs`].
#[allow(dead_code)]
fn server_config_from_pem(cert_pem: &str, key_pem: &str) -> std::io::Result<Arc<ServerConfig>> {
    use std::io::{BufReader, Error, ErrorKind};

    let certs = rustls_pemfile::certs(&mut BufReader::new(std::fs::File::open(cert_pem)?))
        .collect::<std::io::Result<Vec<_>>>()?;
    let key = rustls_pemfile::private_key(&mut BufReader::new(std::fs::File::open(key_pem)?))?
        .ok_or_else(|| Error::new(ErrorKind::InvalidData, "no private key in PEM"))?;

    let cfg = ServerConfig::builder_with_provider(Arc::new(ring::default_provider()))
        .with_safe_default_protocol_versions()
        .map_err(|e| Error::new(ErrorKind::InvalidData, e))?
        .with_no_client_auth()
        .with_single_cert(certs, key)
        .map_err(|e| Error::new(ErrorKind::InvalidData, e))?;
    Ok(Arc::new(cfg))
}
